Kernel/ACPI/ACPIStaticParser.cpp

```diff
@@ -104,26 +105,97 @@ namespace ACPI {
+            auto reset_region = MM.allocate_kernel_region(p_reset_reg.page_base(), (PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
+            klog() << "ACPI: Reboot, Sending value 0x" << String::format("%x", fadt->reset_value) << " to " << p_reset_reg;
+            switch (fadt->reset_reg.access_size) {
+            case (u8)GenericAddressStructure::AccessSize::Byte: {
+                auto* reset_reg = (volatile u8*)reset_region->vaddr().offset(p_reset_reg.offset_in_page().get()).as_ptr();
+                (*reset_reg) = fadt->reset_value;
+                break;
+            }
+            case (u8)GenericAddressStructure::AccessSize::Word: {
+                auto* reset_reg = (volatile u16*)reset_region->vaddr().offset(p_reset_reg.offset_in_page().get()).as_ptr();
+                (*reset_reg) = fadt->reset_value;
+                break;
+            }
+            case (u8)GenericAddressStructure::AccessSize::DWord: {
+                auto* reset_reg = (volatile u32*)reset_region->vaddr().offset(p_reset_reg.offset_in_page().get()).as_ptr();
+                (*reset_reg) = fadt->reset_value;
+                break;
+            }
+            case (u8)GenericAddressStructure::AccessSize::QWord: {
+                auto* reset_reg = (volatile u64*)reset_region->vaddr().offset(p_reset_reg.offset_in_page().get()).as_ptr();
+                (*reset_reg) = fadt->reset_value;
+                break;
+            }
+            default:
+                ASSERT_NOT_REACHED();
+            }
+
+            break;
         }
-        klog() << "ACPI: Reboot, Not supported!";
+        case (u8)GenericAddressStructure::AddressSpace::PCIConfigurationSpace: {
+            // According to the ACPI specification 6.2, page 168, PCI addresses must be confined to devices on Segment group 0, bus 0.
+            auto pci_address = PCI::Address(0, 0, ((fadt->reset_reg.address >> 24) & 0xFF), ((fadt->reset_reg.address >> 16) & 0xFF));
+            klog() << "ACPI: Reboot, Sending value 0x" << String::format("%x", fadt->reset_value) << " to " << pci_address;
+            u32 offset_in_pci_address = fadt->reset_reg.address & 0xFFFF;
+            switch (fadt->reset_reg.access_size) {
+            case (u8)GenericAddressStructure::AccessSize::Byte: {
+                PCI::Access::the().write32_field(pci_address, offset_in_pci_address, fadt->reset_value);
+                break;
+            }
+            case (u8)GenericAddressStructure::AccessSize::Word: {
+                PCI::Access::the().write16_field(pci_address, offset_in_pci_address, fadt->reset_value);
+                break;
+            }
+            case (u8)GenericAddressStructure::AccessSize::DWord: {
+                PCI::Access::the().write8_field(pci_address, offset_in_pci_address, fadt->reset_value);
+                break;
+            }
+            case (u8)GenericAddressStructure::AccessSize::QWord: {
+                ASSERT_NOT_REACHED();
+                break;
+            }
```

Comment:
This code feels very generic and repetitive. Can we make a helper somewhere that takes the access size as an input and calls the correct `writeN_field` based on the access size?
Otherwise this pattern seems like it will repeat again and again, no?

Reply:
I created a method to access generic structures in a more generic fashion.
Anyway, there isn't a simple solution here. ACPI specification 6.2, page 168, states that Access Size field in a `GenericAddressStructure` can be one of the following values:

- Undefined = 0
- Byte access = 1
- Word access = 2
- DWord access = 3
- QWord access = 4

Therefore, there's no easy way to take the access size and put it as an argument for `constexpr sizeof(T) == ...`, because it doesn't align with the number of bytes corresponding to access size.

Reply:
```cpp
size = 1 << (access_size - 1);
```

Reply:
We can do that together with some exceptions :)